### PR TITLE
[WIP] Ensure that random number generation in tests is seeded & standardized

### LINKED ISSUE #2534: non-seeded random numbers in tests
these parts of the unit tests seem to use a random number generator with no
fixed seed, which can cause sporadic test failures (e.g.
https://api.travis-ci.com/v3/job/357101203/log.txt):

```
./image/tests/test_resampling.py:47:    data = np.random.randint(0, 10, shape)
./image/tests/test_resampling.py:223:    data = np.random.randint(0, 10, shape)
./image/tests/test_resampling.py:556:    source_affine[:, 3] = np.append(np.random.randint(0, 4, 3), 1)
./image/tests/test_resampling.py:625:    for theta, phi in np.random.randint(4, size=(5, 2)):
./image/tests/test_resampling.py:731:    x = np.random.random((10,))
./image/tests/test_resampling.py:732:    y = np.random.random((10,))
./image/tests/test_resampling.py:733:    z = np.random.random((10,))
./image/tests/test_image.py:341:    data = np.random.randn(4, 5, 7)
./image/tests/test_image.py:497:    fake_fmri_data = np.random.rand(10, 10, 10, 10)
./image/tests/test_image.py:498:    fake_affine = np.random.rand(4, 4)
./glm/tests/test_contrasts.py:31:    X, Y = np.random.randn(p, q), np.random.randn(p, n)
./glm/tests/test_contrasts.py:41:    X, Y = np.random.randn(p, q), np.random.randn(p, n)
./glm/tests/test_contrasts.py:53:    X, Y = np.random.randn(p, q), np.random.randn(p, n)
./glm/tests/test_contrasts.py:64:    X, Y = np.random.randn(p, q), np.random.randn(p, n)
./glm/tests/test_contrasts.py:93:    X, Y = np.random.randn(p, q), np.random.randn(p, n)
./glm/tests/test_contrasts.py:111:    X, Y = np.random.randn(p, q), np.random.randn(p, n)
./glm/tests/test_contrasts.py:123:    X, Y = np.random.randn(p, q), np.random.randn(p, n)
./glm/tests/test_contrasts.py:142:    X1, V1 = np.random.randn(p), np.ones(p)
./glm/tests/test_second_level.py:100:        X = np.random.randn(p, q)
./glm/tests/test_second_level.py:177:        X = np.random.randn(p, q)
./glm/tests/test_second_level.py:347:        X = pd.DataFrame(np.random.rand(4, 2), columns=['r1', 'r2'])
./glm/tests/test_second_level.py:392:        X = pd.DataFrame(np.random.rand(4, 2), columns=['r1', 'r2'])
./glm/tests/test_first_level.py:221:    X, Y = np.random.randn(p, q), np.random.randn(p, n)
./glm/tests/test_first_level.py:248:    u = np.random.randn(*shape)
./glm/tests/test_first_level.py:249:    mean = 100 * np.random.rand(shape[1]) + 1
./glm/tests/test_thresholding.py:19:    np.random.shuffle(x)
./glm/tests/test_dmtx.py:59:    durations = 5 + 5 * np.random.rand(len(onsets))
./glm/tests/test_dmtx.py:60:    values = 1 + np.random.rand(len(onsets))
./glm/tests/test_dmtx.py:72:    values = 1 + np.random.rand(len(onsets))
./glm/tests/test_dmtx.py:116:    ax = np.random.randn(128, 4)
./glm/tests/test_dmtx.py:121:    ax = np.random.randn(127, 4)
./glm/tests/test_dmtx.py:127:    ax = np.random.randn(128, 4)
./glm/tests/test_dmtx.py:141:    ax = np.random.randn(128, 4)
./glm/tests/test_dmtx.py:345:    ax = np.random.randn(128, 4)
./glm/tests/test_dmtx.py:359:    ax = np.random.randn(128, 4)
./glm/tests/test_dmtx.py:420:    ax = np.random.randn(128, 4)
./glm/tests/test_hemodynamic_models.py:83:    X = np.random.randn(100, 5)
./glm/tests/test_hemodynamic_models.py:92:    X = np.random.randn(100)
./glm/tests/test_utils.py:23:    X = np.random.randn(n, p)
./glm/tests/test_utils.py:33:    p = np.random.rand(10)
./glm/tests/test_utils.py:43:    x = np.random.rand(n) / n
./glm/tests/test_utils.py:44:    A = np.random.rand(n, n) / n
./glm/tests/test_utils.py:52:    x = np.random.randn(n, 3)
./glm/tests/test_utils.py:55:        A = np.random.randn(120, n)
./glm/tests/test_utils.py:58:    i = np.random.randint(3)
./glm/tests/test_utils.py:77:    X = np.random.randn(shape[0], shape[1], shape[2])
./glm/tests/test_paradigm.py:27:    durations = 5 + 5 * np.random.rand(len(onsets))
./glm/tests/test_paradigm.py:28:    values = np.random.rand(len(onsets))
./glm/tests/test_paradigm.py:40:    values = np.random.rand(len(onsets))
./regions/tests/test_region_extractor.py:69:    img = np.zeros((30, 30, 30)) + 0.1 * np.random.randn(30, 30, 30)
./regions/tests/test_region_extractor.py:92:    map_img = np.zeros((30, 30, 30)) + 0.1 * np.random.randn(30, 30, 30)
./plotting/tests/test_img_plotting.py:211:    data = np.random.randn(10, 10, 10)
./plotting/tests/test_img_plotting.py:432:    maps_img = nibabel.Nifti1Image(np.random.random((10, 10, 10)), np.eye(4))
./plotting/tests/test_img_plotting.py:1299:    adjacency_matrix = np.random.random((4, 4))
./input_data/tests/test_nifti_spheres_masker.py:12:    data = np.random.random((3, 3, 3, 5))
./input_data/tests/test_nifti_spheres_masker.py:23:    data = np.random.random((3, 3, 3, 5))
./input_data/tests/test_nifti_spheres_masker.py:48:    data = np.random.random((3, 3, 3, 5))
./input_data/tests/test_nifti_spheres_masker.py:85:    data = np.random.random(shape + (5,))
./input_data/tests/test_nifti_spheres_masker.py:110:    data = np.random.random(shape)
./input_data/tests/test_nifti_spheres_masker.py:141:    data_without_nans = np.random.random((9, 9, 9))
./input_data/tests/test_nifti_spheres_masker.py:161:    data = np.random.random((3, 3, 3, 5))
./input_data/tests/test_nifti_masker.py:185:    data_5d = [np.random.random((10, 10, 10, 1)) for i in range(5)]
./input_data/tests/test_nifti_masker.py:206:    data_5d = [np.random.random((10, 10, 10, 3)) for i in range(5)]
./input_data/tests/test_nifti_masker.py:320:    img = np.random.rand(9, 9, 5)
./input_data/tests/test_nifti_masker.py:400:    signals = np.random.randn(np.prod(data_shape), n_samples)
./input_data/tests/test_nifti_masker.py:401:    means = np.random.randn(np.prod(data_shape), 1) * 50 + 1000
./input_data/tests/test_nifti_maps_masker.py:310:    signals = np.random.randn(np.prod(data_shape), n_samples)
./input_data/tests/test_nifti_maps_masker.py:311:    means = np.random.randn(np.prod(data_shape), 1) * 50 + 1000
./input_data/tests/test_nifti_labels_masker.py:323:    signals = np.random.randn(np.prod(data_shape), n_samples)
./input_data/tests/test_nifti_labels_masker.py:324:    means = np.random.randn(np.prod(data_shape), 1) * 50 + 1000
./input_data/tests/test_multi_nifti_masker.py:166:    imgs = [Nifti1Image(np.random.rand(9, 9, 5), np.eye(4)),
./input_data/tests/test_multi_nifti_masker.py:167:            Nifti1Image(np.random.rand(9, 9, 5), np.eye(4))]
./input_data/tests/test_multi_nifti_masker.py:202:    signals = np.random.randn(2, np.prod(data_shape), n_samples)
./input_data/tests/test_multi_nifti_masker.py:203:    means = np.random.randn(2, np.prod(data_shape), 1) * 50 + 1000
./tests/test_niimg_conversions.py:492:        buffer.append(Nifti1Image(np.random.random((10, 10, 10)), np.eye(4)))
./tests/test_niimg_conversions.py:498:    l = [Nifti1Image(np.random.random((10, 10, 10)), np.eye(4))
./tests/test_masking.py:116:    img1 = Nifti1Image(np.full((9, 9, 9), np.random.rand()), np.eye(4))
./tests/test_masking.py:117:    img2 = Nifti1Image(np.full((9, 9, 9), np.random.rand()), np.eye(4))
./tests/test_masking.py:432:    imgs1 = [Nifti1Image(np.random.randn(9, 9, 9), np.eye(4)),
./tests/test_masking.py:433:             Nifti1Image(np.random.randn(9, 9, 9), np.eye(4))]
./tests/test_masking.py:436:    imgs2 = [Nifti1Image(np.random.randn(9, 9, 9), np.eye(4)),
./tests/test_masking.py:437:             Nifti1Image(np.random.randn(9, 9, 9), np.eye(4))]
./tests/test_niimg.py:63:            dataobj = np.random.uniform(0, 255,
./tests/test_segmentation.py:14:    img = np.zeros((30, 30, 30)) + 0.1 * np.random.randn(30, 30, 30)
./tests/test_segmentation.py:47:    np.random.seed(42)
./tests/test_segmentation.py:48:    img = np.random.normal(size=(3, 3, 3, 3, 3))
./tests/test_segmentation.py:54:    img = np.random.normal(size=(10, 10))
./tests/test_segmentation.py:67:    data = np.zeros((5, 5)) + 0.1 * np.random.randn(5, 5)
./decoding/tests/test_searchlight.py:83:    groups = np.random.permutation(np.arange(frames, dtype=int) >
```

### PR DESCRIPTION
**Pull request by chouhanaryan**
Closes #2534 .

Changes proposed in this pull request:

- Ensure that tests are properly seeded
- Use `RandomState` consistently (as many places as possible except where `sklearn.utils.check_random_state` is used)
- Use as many [Generator](https://numpy.org/doc/stable/reference/random/generator.html) methods with the same name as those of [RandomState](https://numpy.org/doc/stable/reference/random/legacy.html#numpy.random.RandomState) so that in the future if we switch to Generator we can do so easily using a contextual search. The methods replaced are -
    - rand (supported by RandomState only) -> uniform (supported by both Generator and RandomState)
    - randn (supported by RandomState only) -> standard_normal (supported by both Generator and RandomState)

### CODE REVIEW

File: nilearn/input_data/tests/test_nifti_masker.py [hunk 1]
```diff
@@ -203,7 +206,10 @@ def test_5d():
     # Test that, in list of 4d images with last dimension=1, they are
     # considered as 3d
 
-    data_5d = [np.random.random((10, 10, 10, 3)) for i in range(5)]
+    data_5d = [
+        np.random.RandomState(42).random_sample((10, 10, 10, 3))
```
Comment (bthirion) [inline nilearn/input_data/tests/test_nifti_masker.py:210]: Why not creating the generator outside of the loop ?

File: nilearn/tests/test_niimg_conversions.py [hunk 2]
```diff
@@ -489,23 +489,32 @@ def test_concat_niimg_dtype():
...
 
 
 def test_iterator_generator():
     # Create a list of random images
-    l = [Nifti1Image(np.random.random((10, 10, 10)), np.eye(4))
-         for i in range(10)]
-    cc = _utils.concat_niimgs(l)
+    list_images = [
+        Nifti1Image(
+            np.random.RandomState(42).random_sample((10, 10, 10)), np.eye(4)
```
Comment (bthirion) [inline nilearn/tests/test_niimg_conversions.py:505]: Here also I would instantiate the generator outside of the loop.

File: nilearn/tests/test_signal.py [hunk 3]
```diff
@@ -617,14 +617,14 @@ def test_clean_psc():
 
 
 def test_clean_zscore():
-    rng = np.random.RandomState(0)
+    rng = np.random.RandomState(42)
```
Comment (bthirion) [inline nilearn/tests/test_signal.py:620]: When I was 42, I kept wondering what's special with this number --- and did not find.
This is no longer an issue now ;-)

**Review by bthirion — COMMENTED** [review]
Thx, this is great work !
I only have two minor comments.

File: nilearn/input_data/tests/test_nifti_masker.py (hunk 1 above)
Comment (jeromedockes) [reply nilearn/input_data/tests/test_nifti_masker.py:210]: note that as it is, it creates 5 times the same data (the seed is reset every time)

Comment (chouhanaryan) [reply nilearn/input_data/tests/test_nifti_masker.py:210]: Ahha good catch, thanks. Will change it.

**Comment by chouhanaryan** [comment]
As per [this](https://github.com/nilearn/nilearn/issues/2534#issuecomment-699604598) comment in the original issue, we need to document this somewhere too, but I'm not exactly sure where to write what. Let me know and I'll add it in. That should wrap this PR up.

**Comment by bthirion** [comment]
You mean: append the genralization of the use of seeded geenrator in whatsnew.rst ?

**Comment by chouhanaryan** [comment]
I think the contributing documentation might be more appropriate. But again, this seems like a really particular thing to mention. @thomasbazeille any thoughts?

**Comment by thomasbazeille** [comment]
> I think the contributing documentation might be more appropriate. But again, this seems like a really particular thing to mention. @thomasbazeille any thoughts?

Yes indeed but I agree its hard to implement right now given this documentation lacks more major things, drop it for this PR, i'll reference it in the appropriate place (#2142 )

**Comment by chouhanaryan** [comment]
So, is there anything else that's needed for this PR? Just checking in.

File: nilearn/glm/tests/test_second_level.py [hunk 4]
```diff
@@ -344,7 +346,10 @@ def test_second_level_contrast_computation():
             model.compute_contrast(c1, None, None, '')
         # check that passing no explicit contrast when the design
         # matrix has more than one columns raises an error
-        X = pd.DataFrame(np.random.rand(4, 2), columns=['r1', 'r2'])
+        X = pd.DataFrame(
+            np.random.RandomState(42).uniform(size=(4, 2)),
```
Comment (bthirion) [inline nilearn/glm/tests/test_second_level.py:350]: you need to define the generator outside of the loop.

File: nilearn/glm/tests/test_second_level.py [hunk 5]
```diff
@@ -389,7 +394,10 @@ def test_non_parametric_inference_contrast_computation():
             non_parametric_inference(Y, X, [], 'intercept', mask)
         # check that passing no explicit contrast when the design
         # matrix has more than one columns raises an error
-        X = pd.DataFrame(np.random.rand(4, 2), columns=['r1', 'r2'])
+        X = pd.DataFrame(
+            np.random.RandomState(42).uniform(size=(4, 2)),
```
Comment (bthirion) [inline nilearn/glm/tests/test_second_level.py:398]: Here too.

File: nilearn/tests/test_niimg.py [hunk 6]
```diff
@@ -60,8 +60,11 @@ def test_img_data_dtype():
     dtype_matches = []
     with InTemporaryDirectory():
         for logical_dtype in nifti1_dtypes:
-            dataobj = np.random.uniform(0, 255,
-                                        size=(2, 2, 2)).astype(logical_dtype)
+            dataobj = (
+                np.random.RandomState(42)
```
Comment (bthirion) [inline nilearn/tests/test_niimg.py:64]: Here also you need to define
`rng = np.random.RandomState(42)`
outside of the loop, then  `call rng.uniform(0, 255, (2, 2, 2)).astype(logical_dtype)` within the loop.

File: nilearn/tests/test_niimg_conversions.py [hunk 7]
```diff
@@ -489,23 +489,33 @@ def test_concat_niimg_dtype():
 
 def nifti_generator(buffer):
     for i in range(10):
-        buffer.append(Nifti1Image(np.random.random((10, 10, 10)), np.eye(4)))
+        buffer.append(
+            Nifti1Image(
+                np.random.RandomState(42).random_sample((10, 10, 10)),
```
Comment (bthirion) [inline nilearn/tests/test_niimg_conversions.py:494]: again

**Review by bthirion — COMMENTED** [review]
Some small glitches to fix

**Comment by bthirion** [comment]
Thx ! 
The CircleCI failure seems unrelated. Have you rebased on current master ? I think that the sphinx-gallery problem has been fixed in current master.

**Comment by jeromedockes** [comment]
thanks a lot! LGTM

**Comment by chouhanaryan** [comment]
@bthirion rebased!

**Comment by bthirion** [comment]
Merging, thx !

**Comment by chouhanaryan** [comment]
@bthirion Hey, sorry about this but for some reason my PR didn't count towards Hacktoberfest. I suspect this has to do with the new opt-in rules. It'd mean a lot to me if you could add a label to this PR called 'hacktober-accepted'. I'm not sure if it'll work or if it's even possible since it's closed already but could you give it a try? Thanks.

**Comment by bthirion** [comment]
Gael is faster than me...

**Comment by GaelVaroquaux** [comment]
Hi @chouhanaryan, I have added the label.

Tell us if this counted. It was a good PR, with a sizable amount of work. Thank you.

If it did not work, we'll find a way to make it work.

**Comment by chouhanaryan** [comment]
@GaelVaroquaux Thank you! I just realized however that the label needs to be 'hacktoberfest-accepted' (this is just me being inattentive, sorry 🤦‍♂️).

**Comment by GaelVaroquaux** [comment]
I changed the label. Tell us if it works!

**Comment by chouhanaryan** [comment]
It didn't work, unfortunately. It still just says this -

image

I think I'll give it a day or maybe just wait it out until it matures. If it still doesn't count, I'll let you know.
Thank you for all of your help though!

**Comment by GaelVaroquaux** [comment]
OK, ping us back
